pkg/controller/eks/cluster.go

```diff
@@ -188,14 +188,17 @@ func (e *external) Update(ctx context.Context, mg resource.Managed) (managed.Ext
 	if err != nil || rsp.Cluster == nil {
 		return managed.ExternalUpdate{}, errors.Wrap(err, errDescribeFailed)
 	}
+	if add, remove := awsclients.DiffTags(cr.Spec.ForProvider.Tags, rsp.Cluster.Tags); len(remove) != 0 {
+		_, err := e.client.UntagResourceRequest(&awseks.UntagResourceInput{ResourceArn: rsp.Cluster.Arn, TagKeys: remove}).Send(ctx)
+		return managed.ExternalUpdate{}, errors.Wrap(resource.Ignore(eks.IsErrorInUse, err), errAddTagsFailed)
+	} else if len(add) != 0 {
```

Comment:
Is there a specific reason for not calling Tag and Untag consecutively if needed?

Reply:
Mostly because I wasn't sure if these were eventually consistent API calls, in which case I was trying to avoid trying to remove a tag and then trying to add it before it was removed. That being said, if we failed we would just try again so we could likely go ahead and call in succession. For the subsequent update calls, I think it may make sense to issue them on a separate reconcile because if the tags differ that may be the only change required. Alternatively, we could check if the tags were actually changed explicitly in `IsUpToDate`. Do you have a preference?

Reply:
> Mostly because I wasn't sure if these were eventually consistent API calls, in which case I was trying to avoid trying to remove a tag and then trying to add it before it was removed.

I think they block and return only if the tagging is completed. I'm not 100% sure but I haven't seen an AWS call that is not completed immediately after return of the call unless it reports that it's in progress.


I would actually prefer that at the end of the `Update` call, we are on the desired state if possible. In that case, it seems like:
```
add, remove := ...
if err := untag... {}
if err := tag ... {}
```

would work, right? The diff function populates both maps correctly from what I can tell.

Reply:
@muvaf yep! That is correct. Do you think we should also make the other update calls on this reconcile as well?

Reply:
Do you mean not returning after these two calls are made? If so, yes, Update input generators do not include tags so it should be safe but I'm not sure what the API will return if there is no change because the sole reason of `Update` could be the tags. I don't know if that's easily possible but we can either:
* Ignore `no modification requests were made` error which says the update is already the observed state or
* check whether the resulting update input object is different than observation for each call. Something like `EnsureClusterVersion(...)` call where it first checks the diff for the inputs it's looking for and makes the actual `UpdateClusterVersionRequest` call only if there is a diff.

The latter seems safer but I feel like former one has better ROI in terms of complexity because the case where `Update` called but only one of those update calls are necessary happens but happens only once; in the next reconcile `IsUpToDate` returns `true` anyway. But if neither of the approaches are not easy to achieve right now, I'd not persist since even if `no modification requests were made` error is received, the next reconcile will clear that immediately.

Reply:
> But if neither of the approaches are not easy to achieve right now, I'd not persist

Got a little lost in the negation here :sweat_smile: I could modify `IsUpToDate` to return multiple values to say:
- Are tags up to date?
- Is cluster version up to date?
- Is config up to date?

We would trigger `Update` on any of those being `false`, but then only make the individual update requests for the one that is not up to date. Is that aligned with your thinking?

Reply:
@muvaf I opted for always making the cluster config update for now. tag / untag / version update will only happen when needed.